dds/idl/topic_keys.cpp

```diff
@@ -190,17 +191,42 @@ TopicKeys::Iterator::operator++()
       if (!struct_root) {
         throw Error(root_, "Invalid Key Iterator");
       }
-      ACE_CDR::ULong field_count = struct_root->nfields();
+      const ACE_CDR::ULong field_count = struct_root->nfields();
+      // If a nested struct marked as a key has no keys, all the fields are
+      // implied to be keys (expect those marked with @key(FALSE)).
+      if (pos_ == 0) {
+        if (level_ > 0) {
+          implied_keys_ = true;
+          for (; pos_ < field_count && implied_keys_; ++pos_) {
+            AST_Field** field_ptrptr;
+            struct_root->field(field_ptrptr, pos_);
+            AST_Field* field = *field_ptrptr;
```

Comment:
Consider write a helper for this multi-step process of iterating over fields in a struct (since it occurs below as well).  There is already support for something similar in dds_visitor::visit_structure.

Reply:
I'm not sure I understand what you're suggesting here. `dds_visitor::visit_structure` is rewrapping the fields in a vector. It makes sense to avoid using this awkward code in the all generators, but here I would have to either store the vector in the iterator or recreate the vector on each call to `++`. Both of those seem more awkward than direct access, unless I'm not seeing something.

Reply:
I started out thinking that it would be good to have a real iterator type (not necessarily a vector, an iterator that just wraps an integer and has a pointer to the ast node).  In topic_keys, that iterator could be stored in your iterator.  In dds_visitor, the vector constructor that takes a pair of iterators could be used.

But if that's not practical, we could have single helper function like `AST_Field* struct_field(AST_Struct*, int)` could be used by all 3.

Reply:
Okay, I was too transfixed by the idea of using a vector. I will look into creating a simple iterator.

Reply:
Also I forgot about this, but I added an `operator[size_t]` to `AST_Structure` in the original annotation work. It does have the minor problem of returning `AST_Decl*` instead of `AST_Field*` though, which was an oversight on my part.